File: src/1631_Path_With_Minimum_Effort.py
# 1631_Path_With_Minimum_Effort
# https://leetcode.cn/problems/path-with-minimum-effort/description/

# 并查集模板
class UnionFind:
    def __init__(self, n: int):
        # Initialize the parent list where each element is its own parent
        self.parent = list(range(n))
        # Initialize the size list, where each element's size is 1 initially
        self.size = [1] * n
        # Store the number of elements in the UnionFind
        self.n = n
        # Initialize the count of distinct sets
        self.setCount = n
    
    def findset(self, x: int) -> int:
        # Find the root of the set to which 'x' belongs
        # If 'x' is its own parent, it is the root
        if self.parent[x] == x:
            return x
        # Path compression: update the parent of 'x' to its root
        self.parent[x] = self.findset(self.parent[x])
        return self.parent[x]
    
    def unite(self, x: int, y: int) -> bool:
        # Find the roots of the sets to which 'x' and 'y' belong
        x, y = self.findset(x), self.findset(y)
        # If they are already in the same set, return False
        if x == y:
            return False
        # Union by size: make the larger set the parent
        if self.size[x] < self.size[y]:
            x, y = y, x
        # Merge the sets
        self.parent[y] = x
        # Update the size of the root of the merged set
        self.size[x] += self.size[y]
        # Decrease the number of distinct sets
        self.setCount -= 1
        return True
    
    def connected(self, x: int, y: int) -> bool:
        # Check if 'x' and 'y' are in the same set
        x, y = self.findset(x), self.findset(y)
        return x == y

from typing import List

class Solution:
    def minimumEffortPath(self, heights: List[List[int]]) -> int:
        m, n = len(heights), len(heights[0])
        edges = list()
        for i in range(m):
            for j in range(n):
                iden = i * n + j
                if i > 0:
                    edges.append((iden - n, iden, abs(heights[i][j] - heights[i - 1][j])))
                if j > 0:
                    edges.append((iden - 1, iden, abs(heights[i][j] - heights[i][j - 1])))
        
        edges.sort(key=lambda e: e[2])

        uf = UnionFind(m * n)
        ans = 0
        for x, y, v in edges:
            uf.unite(x, y)
            if uf.connected(0, m * n - 1):
                ans = v
                break
        
        return ans
    

File: src/1631_Path_With_Minimum_Effort.rs
// 1631_Path_With_Minimum_Effort
// https://leetcode.cn/problems/path-with-minimum-effort/description/

use std::collections::{HashMap, HashSet};
impl Solution {
    pub fn minimum_effort_path(heights: Vec<Vec<i32>>) -> i32 {
        let mut graph: HashMap<(usize, usize), Vec<((usize, usize), usize)>> = HashMap::new();
        let m = heights.len();
        let n = heights[0].len();
        let mut max_cost = 0;
    
        for i in 0..m {
            for j in 0..n {
                for (dx, dy) in &[(-1, 0), (1, 0), (0, -1), (0, 1)] {
                    let ni = i as i32 + dx;
                    let nj = j as i32 + dy;
                    if ni >= 0 && ni < m as i32 && nj >= 0 && nj < n as i32 {
                        let cost = (heights[i][j] - heights[ni as usize][nj as usize]).abs() as usize;
                        max_cost = max_cost.max(cost);
                        graph.entry((i, j)).or_insert_with(Vec::new).push(((ni as usize, nj as usize), cost));
                    }
                }
            }
        }
    
        fn is_available(
            k: usize,
            m: usize,
            n: usize,
            node: (usize, usize),
            graph: &HashMap<(usize, usize), Vec<((usize, usize), usize)>>,
            visited: &mut HashSet<(usize, usize)>,
            memo: &mut HashMap<((usize, usize), usize), bool>,
        ) -> bool {
            if node == (m - 1, n - 1) {
                return true;
            }
            if let Some(&cached) = memo.get(&(node, k)) {
                return cached;
            }
    
            visited.insert(node);
            if let Some(neighbors) = graph.get(&node) {
                for &(next_node, cost) in neighbors {
                    if cost <= k && !visited.contains(&next_node) {
                        if is_available(k, m, n, next_node, graph, visited, memo) {
                            memo.insert((node, k), true);
                            return true;
                        }
                    }
                }
            }
    
            visited.remove(&node);
            memo.insert((node, k), false);
            false
        }
    
        let mut l = 0;
        let mut r = max_cost;
        let mut visited = HashSet::new();
        let mut memo = HashMap::new();
    
        while l < r {
            let mid = l + (r - l) / 2;
            visited.clear();
    
            if is_available(mid, m, n, (0, 0), &graph, &mut visited, &mut memo) {
                r = mid;
            } else {
                l = mid + 1;
            }
        }
    
        l as i32
    }
}